### DISCUSSION TITLE
Feature: Statistics

### DISCUSSION THREAD

**Discussion by p1xelll** [advplyr/audiobookshelf#167] [Ideas]
I'm a little obsessed with statistics. :) I would welcome some simple stats. Like how much time I spent listening, etc. 

Simple UI Demo:
![Stats - without icon](https://user-images.githubusercontent.com/81173399/140335353-e3b1abf6-2bc4-4560-bdea-d1b3912646c1.png)
[Protype](
https://www.figma.com/proto/0T1L3HPS8iA4S0YLmgJDrK/Untitled?node-id=2%3A2&scaling=min-zoom&page-id=0%3A1&starting-point-node-id=2%3A2) 
[Figma](https://www.figma.com/file/0T1L3HPS8iA4S0YLmgJDrK/Untitled)

Edited: The suggested changes were added to the prototype.

**Comment by juriroemer** [comment]
Overall *minutes* and *size on disk* of books owned would also be interesting! Although that might be more "Database Statistics" or "Server Statistics" than "My Listening Statistics".

**Reply by p1xelll** [reply]
Added total time in hours to the prototype.

**Comment by advplyr** [comment]
This looks great. I really appreciate the mockups, it makes it so much easier to develop new features when I don't need to tinker around with the design.

The introduction of libraries is going to make this page confusing. If this page is only showing stats for the current library, then the minutes listening would need to show only minutes listening for books in that library.
If this page is showing stats across all libraries, then having the stats page in the siderail won't be consistent with the other pages in the siderail.

What do you think? Does that make sense?

**Reply by p1xelll** [reply]
Statistics for the selected library only. I would move the icon for statistics to the top bar.
![Stats - without icon](https://user-images.githubusercontent.com/81173399/140300200-fe2196cc-abd7-4956-9767-1cf03700bc40.png)
I'd leave the side bar just for audiobook related options.

I can also design UI for other features. If you want.

**Reply by advplyr** [reply]
That makes sense.
If it is statistics for the specific library only we may want to change "Books Owned" to "Books in Library" or something.

On the series page I was experimenting with fanning out the series on hover: https://github.com/advplyr/audiobookshelf/discussions/149#discussioncomment-1561488
I'm stuck on whether this should stay and how it should function. Maybe you have some ideas on that?

There is the collections feature, users can create their own collections: https://github.com/advplyr/audiobookshelf/issues/151
I haven't started developing this yet, but there are a lot of design considerations for that.

**Reply by advplyr** [reply]
Oh also, since we are putting the statistics page button in the top right, we don't need to show the siderail. The /config page has its own side navigation and the /upload page doesn't show the siderail.
To remain consistent maybe we don't show the siderail on the /stats page.

**Comment by advplyr** [comment]
I just pushed the first pass of the stats page in `v1.6.12` with the top authors and genres.
Tracking the total time listening is going to be a challenge, so I need to think about that some more.
Let me know what you think so far.

**Reply by p1xelll** [reply]
Even books that are not marked as read are counted as read.

**Reply by advplyr** [reply]
Just fixed in `v1.6.13`

**Comment by advplyr** [comment]
Just pushed `v1.6.16` with the start of recording time listening (experimental).

![image](https://user-images.githubusercontent.com/67830747/141601138-e6b1a8d3-22fd-4869-87dd-54290b3ae34a.png)

In config/users on a user page the browser console will also dump the listening stats for testing

![image](https://user-images.githubusercontent.com/67830747/141601152-bf76eff5-939e-4452-a8ba-79414a2e7f92.png)

If this functions well it will need to be implemented in the app and android auto before it is really useful.

**Comment by advplyr** [comment]
Just released `v1.6.18` where I added the line chart for minutes listening. I had a lot of fun making this from scratch, but when it came to curving the lines I tapped out, too much math for today.

I also realized after I made it that I misinterpreted what your chart was showing. This chart is showing your listening for the last 7 days, but it looks like your chart may have been showing the total listening for each day of the week.

Now that the chart is made, changing the data going into it should be easy. What do you think this chart should be? Maybe it always starts the week on Monday so you can compare weeks (Duolingo app does this).

I don't have much data, The listening data point for today (Sat) was me listening to an audiobook while making this.
![image](https://user-images.githubusercontent.com/67830747/141664128-249113fa-18be-49dd-8825-bbf704729be8.png)

Also updated the top stats
![image](https://user-images.githubusercontent.com/67830747/141664203-89bc9ac7-0bde-4cd3-8765-8f149e2e4659.png)

**Reply by p1xelll** [reply]
The graph shows the total time. I don't know exactly how Duolingo has it, but it would probably be better if people could compare weeks.

**Comment by Maximilian85** [comment]
I think the statistics are very good. I would see as an admin if possible also the summary all users. It is of course not a must but interesting :). What influence have audiobooks that are first downloaded in the app and then listened to?

**Comment by mx03** [comment]
The static works only with books listened with audiobookshelf.
If I mark the books as read because I listend external or before the statistics are wrong. 

Depends on what the statistics should measure if that behavior is right or wrong.

**Reply by advplyr** [reply]
Are you referring specifically to the "Books Read" counter?
When you mark a book as read it is not increasing the counter?

**Reply by mx03** [reply]
![image](https://user-images.githubusercontent.com/1390929/147419493-277bcd37-7439-4ba9-a463-b63c74b10689.png)

If i mark a book as read without read it with audiobookself i listenend that book for my understanding it should count to the total minutes listening time.

But it depends what the entry should mean.

**Reply by advplyr** [reply]
I see what you mean. It is only counting the listening time that it was able to track through the app or web player.
There is a database collection called "sessions" that stores "UserListeningSessions". That number just loops through listening sessions and gets a total.
Maybe it should tally up read books too I'm not sure.

**Comment by advplyr** [comment]
Just released `v1.6.50` with updates to stats.
1. Stats were split into 2 pages - "Library Stats" and "Your Stats" (listening stats)
2. All users now have access to the stats page that shows their listening stats
3. Your Stats page shows the 10 most recent listening sessions
    - a listening session is anytime you start an audiobook stream on web app or mobile
4. New stats on the library stats page including total file size and total audio tracks.

**Comment by argash** [comment]
I just noticed that my stats don't appear to be working. I've listened to the majority of Fallout by Craig Alanson this week. Stats though show almost no listening. I have done most of my listening through mobile safari web app and some on desktop in Brave browser and some through the iOS test flight app. Let me know what to look for in the logs and I can paste whatever is relevant for this.

![image](https://user-images.githubusercontent.com/293324/148769842-fc395a7a-7c37-4f09-ac89-01ffcb40856e.png)

**Reply by advplyr** [reply]
I just saw this. Hopefully the tracking has been better since many updates have been made since you posted this.
I know there is still going to be an issue with listening sessions on the android app for downloaded books.
There is also the listening sessions page where you can see all of your listening sessions with details to narrow down if any of them are wrong/missing.

**Comment by Maximilian85** [comment]
The statistics have really become very extensive and a lot can really be viewed. I don't see it as a priority that the app can't handle downloaded audiobooks yet. Thanks for the great work!

**Comment by skyzuma** [comment]
i would like to see that i can disable the statistics completely so they will no collection of any of these data, i think they pull a lot of performance (but i am not really sure)... i am on a raspberry pi 4 ...

**Comment by SungHerSong** [comment]
Is it possible to edit my listening time? I want to carry over my listening time from the Audible app without leaving audiobooks playing for 35 days straight and messing things up, such as the days listened to, the longest listen, etc.  I would like to add my previous listening time to show up in my stats. I have 35~ days of listening time in the Audible app. The one thing I liked about Audible, was how it tracked my listening time. To the point where I'd play an audiobook on my phone (muted) while listening on another app or website so Audible would update my "master" listening time.

My audiobookshelf setup currently has all of my books from;

Audible .com
Audible .co.uk
Google Books (2 emails)
Authors Direct (Defunct)
Audiobooks .com
YouTube
Author's websites/newsletters/whatever

And a couple of other sites/sources I can't remember.

**Reply by Vito0912** [reply]
I left a comment on Reddit about your question.

**Comment by kujiu** [comment]
I'm testing Audiobookshelf to install it in a very little NGO. We transcribe comics and mangas for blind people, and this tool is what we need. We act in a copyright exemption for disabilities, and we have an agreement to do that. In this case, some statistics are required. We must provide a list of all books, textual or audio, with the number of listening, reading and downloads by month. Is it something possible?

**Reply by Vito0912** [reply]
If you need this to be somewhat official, you can provide that (own work would be needed), but I would strongly discourage using ABS.  
If you are only using the web interface, the stats are mostly accurate. However, if users start using apps, the stats can be incorrect, not tracked at all, or, due to bugs, highly inflated.  
There are no statistics about downloads, although you can check the console for them. Still, some clients do not use the correct endpoints, so these actions will not be logged.  
In theory, you could have all the data, but a user - even by accident - can completely throw off the statistics in either direction.

**Reply by kujiu** [reply]
Thanks
